guides/micronaut-kotlin-extension-fns/micronaut-kotlin-extension-fns.adoc

```diff
@@ -0,0 +1,173 @@
+
+source:DadJokeClient[]
+
+Note that while *we* are creating this client for the purposes of this guide, but often times something like this may come from
+a third party library. Perhaps you're getting it from a public dependency you don't have permission to edit, or from
+another team within your company that has different priorities from your own team.
+
+It could also be that you simply have a specific use case for the client involving specific setup for your application
+that doesn't belong in the client proper.
+
+Let's see how we can extend this client to suit our own use case.
+
+=== DadJokeController
+
+We'll create a controller to utilize the client's standard `GET` for a random joke.
+
+source:DadJokeController[tags=fileHead|start|standardGet|end]
+
+Now, say we have a particular way we want to use a client frequently (like attaching common headers, or filling in some parameters by default.
+
+Let's explore how we can extend the client for our benefit. We'll define a method for the client to specifically look for jokes about dogs.
+
+In your controller file, at the bottom, add this extension function definition:
+
+source:DadJokeController[tag=clientExt]
+
+<1> We define a `getDogJokes` method on the `DadJokeClient`. This method will be available anywhere within the `example.micronaut` package
+
+While this is a simplified (and somewhat silly) example, you can imagine how with a more sophisticated API, this could be
+a very powerful tool to encapsulate common functionality and explicitly relate it to the appropriate class, rather than
+defining another class just to encapsulate this common routine.
+
+We can then use this extension function within our controller by defining a `/dogJoke` endpoint
+
+source:DadJokeController[tag=usingExt]
+
+<1> Note how we can use this function as if it were defined on the client class itself.
+
+Now we have our application!
+
+=== Writing some tests
+
+Lastly, let's use a few more convenient functions included in `micronaut-kotlin-extension-functions` in our test
+
+test:DadJokeTest[]
+
+<1> Here we have `run<EmbeddedServer>` as a little syntatic sugar for `ApplicationContext.run(EmbeddedServer::class.java)`
+<2> Same here for `createBean<HttpClient>`, we're reducing our need to type `::class.java` all over the place
+<3> `retrieveObject` and `retrieveList` give us nice shortcuts to reduce the need for `Argument.of` and `Argument.listOf`, in addition to reducing our `::class.java` uses.
+
+Now we can test everything out!
+
+include::{commondir}/common-testApp.adoc[]
+
+include::{commondir}/common-runapp.adoc[]
+
+include::{commondir}/common-graal-with-plugins.adoc[]
+
+:exclude-for-languages:groovy
+
+Whether you run the app via gradle or as a native image, you should be able to get a good laugh by typing:
+
+`curl localhost:8080/dadJokes/joke`
+
+or
+
+`curl localhost:8080/dadJokes/dogJokes`
+
+Hopefully it brings a smile to your day!
+
+:exclude-for-languages:
```

Comment:
You can probably remove `:exclude-for-languages:groovy` and `:exclude-for-languages:` because this is a Kotlin-only guide.